### PR TITLE
[CSM] Localplayer api lua

### PR DESCRIPTION
**Pull request by Dumbeldor**
#5394

### CODE REVIEW

**Comment by nerzhul** [comment]
You should remove setters at first, we only read interesting values, CSM is not writter on environment at this moment

File: src/script/lua_api/l_localplayer.cpp [hunk 1]
```diff
@@ -0,0 +1,397 @@
...
+
+int LuaLocalPlayer::l_is_climbing(lua_State *L) {
+    return 0;
+}
+
+int LuaLocalPlayer::l_swimming_vertical(lua_State *L) {
+    return 0;
+}
+
+int LuaLocalPlayer::l_get_physics_override_speed(lua_State *L) {
```
Comment (bigfoot547) [inline src/script/lua_api/l_localplayer.cpp]: Maybe use 1 function and return a lua list.

**Comment by bigfoot547** [comment]
Also, please do not use the `ui` namespace for the localplayer reference. Unless I read that wrong and you didn't use that namespace.

**Comment by Dumbeldor** [comment]
Yes i will change the namespace.
Thanks.

File: src/script/lua_api/l_localplayer.cpp [hunk 2]
```diff
@@ -0,0 +1,424 @@
...
+
+int LuaLocalPlayer::l_get_movement_speed_walk(lua_State *L) {
+    LocalPlayer *player = getobject(L, 1);
+
+    lua_pushnumber(L, player->movement_speed_walk);
+    return 1;
+}
+
+int LuaLocalPlayer::l_get_hp(lua_State *L) {
+    LocalPlayer *player = getobject(L, 1);;
```
Comment (nerzhul) [inline src/script/lua_api/l_localplayer.cpp]: many double ;;

**Comment by Dumbeldor** [comment]
It's okay ! 👍

**Comment by sofar** [comment]
Tabs, not spaces, please. Double check the style guide as well.

**Comment by Dumbeldor** [comment]
It's done :)

File: src/script/lua_api/l_localplayer.cpp [hunk 3]
```diff
@@ -0,0 +1,415 @@
...
+}
+
+int LuaLocalPlayer::l_swimming_vertical(lua_State *L) {
+	LocalPlayer *player = getobject(L, 1);
+
+	lua_pushboolean(L, player->swimming_vertical);
+	return 1;
+}
+
+int LuaLocalPlayer::l_get_physics_override_speed(lua_State *L) {
```
Comment (bigfoot547) [inline src/script/lua_api/l_localplayer.cpp]: Please use 1 function and return a lua list with the syntax:
`{speed = <number>, jump = <number>, gravity = <number>, sneak = <bool>, sneak_glitch = <bool>}`.

**Comment by kaeza** [comment]
Docs are missing, and please, can you make it as similar as possible to the `ObjectRef` API? Otherwise modders will have to remember yet another set of functions. E.g. `getvelocity` instead of `get_speed`..

File: src/script/lua_api/l_localplayer.cpp (hunk 3 above)
Comment (nerzhul) [inline src/script/lua_api/l_localplayer.cpp]: @bigfoot547 you are now clear, you mean merge all physics on a get_physics_override() ?

**Comment by nerzhul** [comment]
@kaeza i think he uses calls due to engine attributes, hopefully this can be fixed :)

File: src/script/lua_api/l_localplayer.cpp (hunk 3 above)
Comment (Dumbeldor) [inline src/script/lua_api/l_localplayer.cpp]: @bigfoot547 I don't know.
I don't like send things that is useless.

**Comment by Dumbeldor** [comment]
Yeah @nerzhul 
@kaeza i change that, thanks.

File: src/script/lua_api/l_localplayer.cpp (hunk 3 above)
Comment (bigfoot547) [inline src/script/lua_api/l_localplayer.cpp]: @Dumbeldor That makes sense, but when I call `get_physics_override()`, I usually use `speed` and `jump`. Making 2 function calls feels inefficient. Also, keeping similar syntax with server-side mods makes it so you don't need to learn 2 very different APIs.

Comment (SmallJoker) [inline src/script/lua_api/l_localplayer.cpp]: I agree that these parameters should be bundled together. It reduces the function count dramatically and easily allows adding more parameters to the returned table.

Comment (Dumbeldor) [inline src/script/lua_api/l_localplayer.cpp]: Okay i understand 👍

File: src/client.cpp [hunk 4]
```diff
@@ -1864,6 +1864,7 @@ void Client::afterContentReceived(IrrlichtDevice *device)
 
 	m_state = LC_Ready;
 	sendReady();
+	m_script->on_client_ready(m_env.getLocalPlayer());
```
Comment (nerzhul) [inline src/client.cpp]: you should verify if enable_client_modding setting is true

File: src/script/lua_api/l_localplayer.h [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+	static int l_get_movement_acceleration_fast(lua_State *L);
+	static int l_get_movement_speed_walk(lua_State *L);
+	static int l_get_movement_speed_crouch(lua_State *L);
+	static int l_get_movement_speed_fast(lua_State *L);
+	static int l_get_movement_speed_climb(lua_State *L);
+	static int l_get_movement_speed_jump(lua_State *L);
+	static int l_get_movement_liquid_fluidity(lua_State *L);
+	static int l_get_movement_liquid_fluidity_smooth(lua_State *L);
+	static int l_get_movement_liquid_sink(lua_State *L);
+	static int l_get_movement_gravity(lua_State *L);
```
Comment (red-001) [inline src/script/lua_api/l_localplayer.h]: This could be merged into a single function.

File: src/script/lua_api/l_localplayer.cpp [hunk 6]
```diff
@@ -0,0 +1,410 @@
...
+	// Keep localplayer object stack id
+	int localplayer_object = lua_gettop(L);
+
+	lua_getglobal(L, "core");
+	lua_getfield(L, -1, "localplayer");
+	luaL_checktype(L, -1, LUA_TTABLE);
+	int uitable = lua_gettop(L);
+
+	lua_pushvalue(L, localplayer_object);
+	lua_setfield(L, uitable, "localplayer");
```
Comment (red-001) [inline src/script/lua_api/l_localplayer.cpp]: Not sure if I'm reading this correctly but doesn't this make the object accessible at `core.localplayer.localplayer`? That doesn't seem like a good location for it.

File: src/script/clientscripting.cpp [hunk 7]
```diff
@@ -68,4 +72,11 @@ void ClientScripting::InitializeModApi(lua_State *L, int top)
 	LuaItemStack::Register(L);
 	StorageRef::Register(L);
 	LuaMinimap::Register(L);
+	LuaLocalPlayer::Register(L);
+}
+
+void ClientScripting::on_client_ready(LocalPlayer *localplayer)
```
Comment (red-001) [inline src/script/clientscripting.cpp]: When you add documentation you should add a warning that the localplayer reference isn't always accessible.

File: src/script/lua_api/l_localplayer.cpp (hunk 6 above)
Comment (Dumbeldor) [inline src/script/lua_api/l_localplayer.cpp]: Yes you are right.
I change to core.localplayer

**Comment by Dumbeldor** [comment]
Begin doc, i continue tomorrow ;)

File: src/script/clientscripting.cpp [hunk 8]
```diff
@@ -45,6 +46,9 @@ ClientScripting::ClientScripting(Client *client):
 	lua_newtable(L);
 	lua_setfield(L, -2, "ui");
 
+	lua_newtable(L);
+	lua_setfield(L, -2, "localplayer");
+
```
Comment (red-001) [inline src/script/clientscripting.cpp]: Is this still needed?

File: src/script/lua_api/l_localplayer.cpp [hunk 9]
```diff
@@ -0,0 +1,377 @@
...
+}
+
+int LuaLocalPlayer::l_get_last_speed(lua_State *L) {
+	LocalPlayer *player = getobject(L, 1);
+
+	push_v3f(L, player->last_speed);
+	return 1;
+}
+
+int LuaLocalPlayer::l_get_last_pitch(lua_State *L) {
```
Comment (red-001) [inline src/script/lua_api/l_localplayer.cpp]: please rename this to `l_get_last_look_vertical`

File: src/script/lua_api/l_localplayer.cpp [hunk 10]
```diff
@@ -0,0 +1,377 @@
...
+	LocalPlayer *player = getobject(L, 1);
+
+	lua_pushnumber(L, player->last_pitch);
+	return 1;
+}
+
+int LuaLocalPlayer::l_get_last_yaw(lua_State *L) {
```
Comment (red-001) [inline src/script/lua_api/l_localplayer.cpp]: please rename this to `l_get_last_look_horizontal`

File: src/script/clientscripting.cpp (hunk 8 above)
Comment (Dumbeldor) [inline src/script/clientscripting.cpp]: I forgot to remove it.

File: doc/client_lua_api.md [hunk 11]
```diff
@@ -774,6 +774,37 @@ An interface to manipulate minimap on client UI
...
+* `is_in_liquid_stable()`:
+* `get_liquid_viscosity()`:
+* `is_climbing()`:
+* `swimming_vertical()`:
+* `get_physics_override()`:
+* `get_override_position()`:
+* `get_last_position()`:
+* `get_last_speed()`:
+* `get_last_pitch()`:
+* `get_last_yam()`:
```
Comment (bigfoot547) [inline doc/client_lua_api.md]: Nope, no yams here! :smile:
Please rename to `get_last_horizontal`.

File: doc/client_lua_api.md [hunk 12]
```diff
@@ -774,6 +774,37 @@ An interface to manipulate minimap on client UI
 * `get_mode()`: returns the current minimap mode
 * `toggle_shape()`: toggles minimap shape to round or square.
 
+### LocalPlayer
```
Comment (bigfoot547) [inline doc/client_lua_api.md]: Maybe tell how to access the object. `core.localplayer` right?

File: doc/client_lua_api.md [hunk 13]
```diff
@@ -774,6 +774,37 @@ An interface to manipulate minimap on client UI
...
+An interface to retrieve information about the player.
+
+* `get_velocity()`: returns player send
```
Comment (bigfoot547) [inline doc/client_lua_api.md]: I am not understanding something...

Comment (Dumbeldor) [inline doc/client_lua_api.md]: "returns player speed" *

File: doc/client_lua_api.md (hunk 11 above)
Comment (DS-Minetest) [inline doc/client_lua_api.md]: If you call it `get_last_horizontal` (or `get_last_look_horizontal`), the `get_last_pitch()` also has to be changed to `get_last_vertical` (or `get_last_look_vertical`). If you leave the `get_last_pitch()` as it is, call `get_last_yam()` `get_last_yaw()`.

File: src/script/lua_api/l_localplayer.cpp (hunk 9 above)
Comment (Dumbeldor) [inline src/script/lua_api/l_localplayer.cpp]: @rubenwardy it's depretaced

File: doc/client_lua_api.md (hunk 13 above)
Comment (bigfoot547) [inline doc/client_lua_api.md]: Okay, that makes sense. Thanks!

File: doc/client_lua_api.md (hunk 12 above)
Comment (bigfoot547) [inline doc/client_lua_api.md]: Right???

Comment (red-001) [inline doc/client_lua_api.md]: seems so

File: src/script/lua_api/l_localplayer.cpp [hunk 14]
```diff
@@ -0,0 +1,377 @@
...
+You should have received a copy of the GNU Lesser General Public License along
+with this program; if not, write to the Free Software Foundation, Inc.,
+51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+*/
+
+#include "script/common/c_converter.h"
+#include "l_localplayer.h"
+#include "l_internal.h"
+
+LuaLocalPlayer::LuaLocalPlayer(LocalPlayer *m) {
```
Comment (SmallJoker) [inline src/script/lua_api/l_localplayer.cpp]: Mass-replacement required here. Please move the bracket to a new line, as described in our [code style guidelines](http://dev.minetest.net/Code_style_guidelines).

File: src/script/lua_api/l_localplayer.cpp [hunk 15]
```diff
@@ -0,0 +1,377 @@
...
+
+
+	push_v3f(L, player->getSpeed() / BS);
+	return 1;
+}
+
+int LuaLocalPlayer::l_get_hp(lua_State *L) {
+	LocalPlayer *player = getobject(L, 1);
+
+	lua_pushinteger(L, player->hp);
```
Comment (SmallJoker) [inline src/script/lua_api/l_localplayer.cpp]: All these functions are not fail-safe.
i.e. `core.localplayer.get_hp()` triggers an access violation crash. (correct: `core.localplayer:get_hp()`)

Comment (bigfoot547) [inline src/script/lua_api/l_localplayer.cpp]: Yeah, I'd use this:
```c++
if(!player) {
	return 0;
}
```

Comment (nerzhul) [inline src/script/lua_api/l_localplayer.cpp]: @SmallJoker we are in lua , it's a static function, we don't have access to member then

File: doc/client_lua_api.md [hunk 16]
```diff
@@ -774,6 +774,37 @@ An interface to manipulate minimap on client UI
...
+* `get_last_position()`:
+* `get_last_speed()`:
+* `get_last_pitch()`:
+* `get_last_yam()`:
+* `get_last_key_pressed()`:
+* `get_breath()`:
+* `get_look_dir()`:
+* `get_look_horizontal()`:
+* `get_look_vertical()`:
+* `get_position()`:
```
Comment (rubenwardy) [inline doc/client_lua_api.md]: `get_pos`

**Comment by Dumbeldor** [comment]
@rubenwardy No, not possible to reuse server SAO.
For the moment we want the local player.

**Comment by rubenwardy** [comment]
I don't believe that. Surely the SAO and CAO both inherit from the same parent

**Comment by nerzhul** [comment]
Yes from activeobject which is tiny and doesn't have bindings

**Comment by bigfoot547** [comment]
*But*, we are using the `LocalPlayer` class right? Not `GenericCAO`?

**Comment by Dumbeldor** [comment]
I finished the documentation. But I am French and my level of English is not very good yet. I'm waiting for your review! :)

File: doc/client_lua_api.md (hunk 12 above)
Comment (Dumbeldor) [inline doc/client_lua_api.md]: No, I do not think so. For the minimap nrzkt just wrote "Minimap"

**Comment by rubenwardy** [comment]
> But, we are using the LocalPlayer class right? Not GenericCAO?

Oh I see. LocalPlayer inherits from Player, like RemotePlayer does, although it may be difficult given that I believe that a playerref is just a luaentity ref

**Comment by nerzhul** [comment]
@rubenwardy yeah, LocalPlayer is not PlayerCAO object, it's a little bit less clean than server on the client side

File: src/script/lua_api/l_localplayer.cpp [hunk 17]
```diff
@@ -0,0 +1,408 @@
...
+
+	lua_pushvalue(L, localplayer_object);
+	lua_setfield(L, uitable, "localplayer");
+}
+
+int LuaLocalPlayer::l_get_velocity(lua_State *L)
+{
+	LocalPlayer *player = getobject(L, 1);
+
+
```
Comment (rubenwardy) [inline src/script/lua_api/l_localplayer.cpp]: unnecessary space

File: doc/client_lua_api.md (hunk 12 above)
Comment (rubenwardy) [inline doc/client_lua_api.md]: you should document how to get it here as well. Don't necessary take examples from the rest of the docs

**Comment by rubenwardy** [comment]
The way you get it from minetest.localplayer worries me, you have to be absolutely sure that it's delete on shutdown if necessary before any on_shutdown calls. I'd prefer a minetest.get_local_player() in this respect.

**Comment by Dumbeldor** [comment]
@rubenwardy it is useless. He can store the localplayer in lua variable. :p

**Comment by nerzhul** [comment]
@rubenwardy it doesn't change any thing if you store the player lua object in another variable and on_shutdown is called. Localplayer should be emptied at the lifetime end of the CPP object, @Dumbeldor can you tell us where LocalPlayer CPP object is destroyed ? And if it's before on_shutdown invalidate core.localplayer variable ?

**Comment by nerzhul** [comment]
@rubenwardy in fact i found a memleak, we never remove LocalPlayer object.

LocalPlayer is created with ClientEnvironment in Client constructor. Lua stack is created after that, but we link localplayer to lua stack when LocalPlayer is consistent (data loaded from server).
LuaStack is destroyed on Client::Stop(). We will remove LocalPlayer when ClientEnvironment will be destroyed, and it's destroyed when Client is destroy, then after Client::stop().

Creation lifecycle: 

```
Client -> ClientEnv -> LocalPlayer -> luastack -> connection -> localplayer to lua
```
Destruction lifecycle

```
luastack > LocalPlayer -> ClientEnv -> Client
```

**Comment by nerzhul** [comment]
The LocalPlayer memleak is fixed https://github.com/minetest/minetest/commit/54ffc4e915fcdb6f6e7556bc2d1f8b27fd567d9f

**Comment by Dumbeldor** [comment]
thank you ner'zhul. I had seen that the destructor was never called. : P

File: doc/client_lua_api.md [hunk 18]
```diff
@@ -809,14 +810,15 @@ An interface to retrieve information about the player.
 * `swimming_vertical()`: returns true if player swimming in vertical
 * `get_physics_override()`: returns :
 
-
+```
```
Comment (bigfoot547) [inline doc/client_lua_api.md]: Now I am being particular, but maybe you could use:
` ```lua ` for syntax highlighting.

**Comment by sofar** [comment]
Unfortunately this needs a rebase!

**Comment by sofar** [comment]
At 30 commits, this really needs to be squashed, too.

**Comment by sofar** [comment]
Core looks OK otherwise, I tested a few things.

 - There's no `get_name()`. Any reason for that?
 - it takes 2-3 seconds before `core.localplayer` returns a valid handle. Any idea how to reliably get it? Running `minetest.after` in a loop works, of course.

File: src/script/lua_api/l_localplayer.cpp [hunk 19]
```diff
@@ -0,0 +1,407 @@
...
+	*(void **) (lua_newuserdata(L, sizeof(void *))) = o;
+	luaL_getmetatable(L, className);
+	lua_setmetatable(L, -2);
+
+	// Keep localplayer object stack id
+	int localplayer_object = lua_gettop(L);
+
+	lua_getglobal(L, "core");
+	luaL_checktype(L, -1, LUA_TTABLE);
+	int uitable = lua_gettop(L);
```
Comment (nerzhul) [inline src/script/lua_api/l_localplayer.cpp]: can you rename uitable to more reliable name ?

**Comment by Dumbeldor** [comment]
@sofar You must wait until LocalPlayer is created. It is created after loading LocalPlayer data from the server. And thank you I added get_name() :)

**Comment by sofar** [comment]
Thanks for squashing.

The documentation style is pretty mis-matched with the rest. Try and format the members just like all the minetest.* members are formatted, with one line for the function name and args, and one or more lines for the explanation and arg breakout, as well as return values.

**Comment by sofar** [comment]
Note that I still have issues getting the localplayer handle in a test mod, sometimes entirely. I consider that a major usability issue.

**Comment by nerzhul** [comment]
@sofar please finish the review permitting merge. For the localplayer handle don't try to obtain it before on_connect event

**Comment by sofar** [comment]
Docs look OK now (repetitive, but properly formatted and reads nicely now)

**Comment by nerzhul** [comment]
sofar okay for merging now ?

**Comment by sofar** [comment]
I had never heard of the `on_connect()` event. If that solves the problem then I think that the proper method to obtain the localplayer handle should be diocumented **varbatim and in full** in the documentation file.

So, where is this minetest.register_on_connect() event documented?

**Comment by Dumbeldor** [comment]
@sofar Yeah, it did not exist. I added the event => #5540

**Comment by sofar** [comment]
Documentation is all wrong now. It says `minetest.get_pos()` but the method is part of `LocalPlayer` and minetest.get_pos() will throw an exception.

**Comment by sofar** [comment]
Good thing is that the `on_connect()` method actually worked:

```
local me

local function logpos()
	print(minetest.pos_to_string(me:get_pos()))
	minetest.after(1, logpos)
end

core.register_on_connect(function()
	me = core.localplayer
	logpos()
end)
```

**Comment by sofar** [comment]
Bad thing is that `get_pos()` exposes BS positions!!!

EXPECTED BEHAVIOR:
1. Turn on noclip so you don't fall
2. type `/teleport 10 10 10
3. above lua code prints `(10, 10, 10)` or similar

ACTUAL BEHAVIOR:
3. lua code prints `(100.00000762939,100.00000762939,100.00000762939)`

**Comment by sofar** [comment]
also, note this patch and the `on_connect()` patch conflict, and both are needed at the same time to test. Perhaps the on_connect() patch should just be part of this PR thread. It's certainly small enough.

File: src/script/lua_api/l_localplayer.h [hunk 20]
```diff
@@ -0,0 +1,95 @@
...
+#include "l_base.h"
+
+
+class LocalPlayer;
+
+class LuaLocalPlayer : public ModApiBase {
+private:
+
+	static const char className[];
+	static const luaL_reg methods[];
```
Comment (nerzhul) [inline src/script/lua_api/l_localplayer.h]: luaL_Reg please

**Comment by nerzhul** [comment]
@sofar i merged the on_connect, fix this PR conflict and fix positions. Please note sofar you can edit PR sometimes, when it doesn't take many time it can be good to collaborate with PR owners, especially when you want it :p

**Comment by Dumbeldor** [comment]
@sofar I don't have to divide the position by BS?

**Comment by nerzhul** [comment]
You should and i fixed it before your squash

**Comment by Dumbeldor** [comment]
@nerzhul Oh okay, thanks ! :)

File: doc/client_lua_api.md (hunk 12 above)
Comment (sofar) [inline doc/client_lua_api.md]: See below, I've put the *exact* code in the docs.

**Comment by sofar** [comment]
I've added a commit (to be squashed) that fixes the documentation up exactly as I would want it to be fied.

**Comment by sofar** [comment]
@rubenwardy is this ok now with you? Your review still shows as open.

**Comment by sofar** [comment]
8ad3dad1

**Pull request closed without merging**